File: 프로그래머스/3/77886. 110 옮기기/110 옮기기.swift
import Foundation

func solution(_ s: [String]) -> [String] {
    var answer: [String] = []
    answer.reserveCapacity(s.count)

    let one: UInt8 = 49  // '1'
    let zero: UInt8 = 48 // '0'

    for str in s {
        var stack: [UInt8] = []
        stack.reserveCapacity(str.utf8.count)

        var cnt110 = 0

        // 1) "110"을 스택으로 모두 제거하며 개수 카운트
        for b in str.utf8 {
            stack.append(b)
            if stack.count >= 3 {
                let n = stack.count
                if stack[n - 3] == one && stack[n - 2] == one && stack[n - 1] == zero {
                    stack.removeLast(3)
                    cnt110 += 1
                }
            }
        }

        // 2) 삽입 위치: 남은 문자열에서 마지막 '0' 뒤 (없으면 0)
        var insertIdx = 0
        if !stack.isEmpty {
            for i in stride(from: stack.count - 1, through: 0, by: -1) {
                if stack[i] == zero {
                    insertIdx = i + 1
                    break
                }
            }
        }

        // 3) 결과 조립: prefix + ("110"*cnt) + suffix
        var res: [UInt8] = []
        res.reserveCapacity(stack.count + cnt110 * 3)

        if insertIdx > 0 {
            res.append(contentsOf: stack[0..<insertIdx])
        }

        if cnt110 > 0 {
            for _ in 0..<cnt110 {
                res.append(one); res.append(one); res.append(zero)
            }
        }

        if insertIdx < stack.count {
            res.append(contentsOf: stack[insertIdx..<stack.count])
        }

        answer.append(String(decoding: res, as: UTF8.self))
    }

    return answer
}

File: 프로그래머스/3/77886. 110 옮기기/110 옮기기.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

// s_len은 배열 s의 길이입니다.
// 파라미터로 주어지는 문자열은 const로 주어집니다. 변경하려면 문자열을 복사해서 사용하세요.
char** solution(const char* s[], size_t s_len) {
    char** answer = (char**)malloc(sizeof(char*) * s_len);

    for (size_t idx = 0; idx < s_len; idx++) {
        const char* str = s[idx];
        int n = (int)strlen(str);

        // 스택(남는 문자들)
        char* stack = (char*)malloc((size_t)n + 1);
        int top = 0;

        int cnt110 = 0;

        // 1) "110" 제거(스택 사용)
        for (int i = 0; i < n; i++) {
            stack[top++] = str[i];

            if (top >= 3 &&
                stack[top - 3] == '1' &&
                stack[top - 2] == '1' &&
                stack[top - 1] == '0') {
                top -= 3;
                cnt110++;
            }
        }

        // 2) 삽입 위치: 남은 문자열(stack[0..top-1])에서 마지막 '0' 뒤
        int insertIdx = 0; // 기본: 맨 앞
        for (int i = top - 1; i >= 0; i--) {
            if (stack[i] == '0') {
                insertIdx = i + 1;
                break;
            }
        }

        // 3) 결과 조립: prefix + ("110"*cnt110) + suffix
        int resLen = top + cnt110 * 3;
        char* res = (char*)malloc((size_t)resLen + 1);

        int p = 0;

        // prefix
        if (insertIdx > 0) {
            memcpy(res + p, stack, (size_t)insertIdx);
            p += insertIdx;
        }

        // "110" 반복 삽입
        for (int k = 0; k < cnt110; k++) {
            res[p++] = '1';
            res[p++] = '1';
            res[p++] = '0';
        }

        // suffix
        if (insertIdx < top) {
            memcpy(res + p, stack + insertIdx, (size_t)(top - insertIdx));
            p += (top - insertIdx);
        }

        res[p] = '\0';

        free(stack);
        answer[idx] = res;
    }

    return answer;
}